// Copyright 2015 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

import "google/protobuf/timestamp.proto";

package discount;

// The Discount service definition.
service Discounter {
  // Calculate discount by product
  rpc calculate_discount (DiscountRequest) returns (DiscountResponse) {}
}

message Discount {
  float pct = 1;
  int32 value_in_cents = 2;
}

message Product {
  string id = 1;
  int32 price_in_cents = 2;
  string title = 3;
  string description = 4;
  Discount discount = 5;

}

message User {
  string id = 1;
  string first_name = 2;
  string last_name = 3;
  google.protobuf.Timestamp date_of_birth = 4;
}

// The request message contains the product and user identifiers
message DiscountRequest {
  string product_id = 1;
  string user_id = 2;
}

// The response message contains the
message DiscountResponse {
  string product_id = 1;
  int32 full_price_in_cents = 2;
  int32 discounted_price_in_cents = 3;
}